### PR TITLE
Access logging and revoking access

### PR DESCRIPTION
**Pull request by GDay**
- Logging what manual integration items have been enabled
- Adding the option to revoke access
- Adding the option to create empty user access integrations. These integrations are meant to be used when there is no api available. An admin will then mark it enabled manually and when offboarding sequences have been implemented they will get a notification to revoke the access manually.

Fixes: https://github.com/chiefonboarding/ChiefOnboarding/issues/315

Example (`revoke` block):

```
{
    "form": [
        {
            "id": "TEAM_ID",
            "url": "https://app.asana.com/api/1.0/organizations/{{ORG}}/teams",
            "name": "Select team to add user to",
            "type": "choice",
            "method": "GET",
            "data_from": "data",
            "choice_name": "name",
            "choice_value": "gid"
        }
    ],
    "exists": {
        "url": "https://app.asana.com/api/1.0/users/{{email}}",
        "method": "GET",
        "expected": "\"{{email}}\"",
        "fail_when_4xx_response_code": false
    },
    "revoke": [
        {
            "url": "https://app.asana.com/api/1.0/workspaces/{{ORG}}/removeUser",
            "data": {
                "data": {
                    "user": "{{email}}"
                }
            },
            "method": "POST"
        }
    ],
    "execute": [
        {
            "url": "https://app.asana.com/api/1.0/workspaces/{{ORG}}/adUser",
            "data": {
                "data": {
                    "user": "{{email}}"
                }
            },
            "method": "POST"
        },
        {
            "url": "https://app.asana.com/api/1.0/teams/{{TEAM_ID}}/addUser",
            "data": {
                "data": {
                    "user": "{{email}}"
                }
            },
            "method": "POST"
        }
    ],
    "headers": {
        "Accept": "application/json",
        "Content-Type": "application/json",
        "Authorization": "Bearer {{TOKEN}}"
    },
    "initial_data_form": [
        {
            "id": "TOKEN",
            "name": "Please put your token here",
            "description": "You can find your token here: https://...."
        },
        {
            "id": "ORG",
            "name": "Organization id",
            "description": "You can find your organization id here: https://..."
        }
    ]
}
```

### CODE REVIEW

**Comment by cscheng** [comment]
My initial thought was that revoking users could already be accomplished using existing functionality, for instance by creating custom integrations "Create 3rd party service account" and "Remove 3rd party service account" (and simply log when a certain integration ran for which user). However, having the `revoke` property in the manifest makes the integration feel more "whole" and it makes sense to have the external API create/delete methods in the same integration/manifest instead of having separate ones.

I would suggest to keep the `revoke` value similar to `execute`. The `verify` option seems a bit redundant, as you expect when the API call is successful the intended action has happened. However, you do need to handle the case when the 3rd party account has been deleted manually before you run this integration. Perhaps this was the intent of `verify`? Customizing the success message also seems a bit gratuitous, you could create a generic message based on the name of the integration.

In absence of an offboarding sequence, using empty custom integrations to notify an admin to manually remove/disable an account feels like a hack. I understand the need to log when a 3rd party account has been created or deleted, but the validity of the data can be questioned as it has be to input manually. Would an admin task not suffice? If offboarding sequences are off the table in the foreseeable future, could there be a more elegant solution than empty custom integrations?

I'm also interested in how these logs are made available to the admins to view/use. Does it need to be exported for audit purposes?

**Comment by GDay** [comment]
Thanks for your thoughts. Definitely some good points.

> I would suggest to keep the revoke value similar to execute. The verify option seems a bit redundant, as you expect when the API call is successful the intended action has happened. However, you do need to handle the case when the 3rd party account has been deleted manually before you run this integration. Perhaps this was the intent of verify? Customizing the success message also seems a bit gratuitous, you could create a generic message based on the name of the integration.

When an admin goes to the new hire -> access part, it will fetch the access status live for each integration. Meaning, if someone revoked access through a different way, then it will not show as active anymore.

> The verify option seems a bit redundant, as you expect when the API call is successful the intended action has happened.

I added `verify` for the edge cases in which maybe an API has changed and suddenly needs another call for an account to be fully removed. Though, I think this is rare enough that we can ignore it for now. Worst case, I can just make it mandatory to check before completing the revoke action. Good call.

> In absence of an offboarding sequence

Yeah, I should have provided a bit more context. Offboarding sequences are on the list and will be done soon!

> Does giving a laptop to a new hire require the same logging?

Hardware tracking is on the list too! 

> I'm also interested in how these logs are made available to the admins to view/use. Does it need to be exported for audit purposes?

Yeah, again, I should have provided more context. One of the features that will be created is the ability to use the integrations data in the integrations (just like we can use {{email}} etc) to show what integrations have been disabled for the new hire. This info is needed for one API endpoint. For example: create a document with what a user had access to and what has been revoked. So we indeed need to track the "dummy" ones. These will also be added to sequences to be created and also be removed (offboarding).

File: back/admin/people/templates/new_hire_access.html [hunk 1]
```diff
@@ -8,33 +8,19 @@
...
+      {% include "_user_access_card.html" %}
+    </div>
+    {% endfor %}
   </div>
-  {% empty %}
-  <div class="card">
+{% else %}
+  <div class="card mt-3">
     <div class="card-body">
       {% translate "You have not enabled any integrations for account provisitioning." %}
```
Comment (cscheng) [inline back/admin/people/templates/new_hire_access.html:22]: Typo: "provisitioning"

File: back/admin/people/templates/colleague_access.html [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    {% for integration in integrations %}
+    <div class="column col-3 integration-{{integration.id}} mb-2" hx-trigger="load" hx-get="{% url 'people:user_check_integration' object.id integration.id %}">
+      {% include "_user_access_card.html" %}
+    </div>
+    {% endfor %}
+  </div>
+{% else %}
+  <div class="card mt-3">
+    <div class="card-body">
+      {% translate "You have not enabled any integrations for account provisitioning." %}
```
Comment (cscheng) [inline back/admin/people/templates/colleague_access.html:17]: Typo: "provisitioning"

File: back/admin/people/access_views.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        return redirect("people:delete", user.id)
+
+
+class UserCheckAccessView(LoginRequiredMixin, IsAdminOrNewHireManagerMixin, DetailView):
+    template_name = "_user_access_card.html"
+    model = get_user_model()
+
+    def get_template_names(self):
+        if "compact" in self.request.path:
```
Comment (cscheng) [inline back/admin/people/access_views.py:90]: "compact" variant is not tested

**Review by cscheng — APPROVED** [review]

File: back/admin/integrations/serializers.py [hunk 4]
```diff
@@ -85,6 +85,20 @@ def validate(self, data):
         return data
 
 
+class ManifestRevokeSerializer(ValidateMixin, serializers.Serializer):
+    url = serializers.CharField()
+    data = serializers.JSONField(required=False, default=dict)
+    headers = serializers.DictField(child=serializers.CharField(), default=dict)
+    method = serializers.ChoiceField(
+        [
+            ("HEAD", "HEAD"),
```
[the remark this answers is not in the document]
Comment (GDay) [reply back/admin/integrations/serializers.py:94]: HEAD is indeed not necessary. Some APIs might need a GET request to remove a user. For example: we would first need to lookup a user with their email address (GET request) to get a user ID and then the second request would be the remove request using the previously fetched ID.

File: back/admin/integrations/models.py [hunk 5]
```diff
@@ -119,13 +128,17 @@ class ManifestType(models.IntegerChoices):
...
+        return self.manifest_type == Integration.ManifestType.MANUAL_USER_PROVISIONING
+
     @property
     def schedule_name(self):
         return f"User sync for integration: {self.id}"
 
     def clean(self):
-        if not self.manifest:
-            # ignore field if form doesn't have it
+        if not self.manifest or self.skip_user_provisioning:
```
[the remark this answers is not in the document]
Comment (GDay) [reply back/admin/integrations/models.py:140]: Yeah, it seems like we don't need the check, but we actually do need it to setup the Google login. The form there, doesn't have the `manifest` field and therefore fails. See: https://github.com/chiefonboarding/ChiefOnboarding/actions/runs/6698961561/job/18202149900

**Pull request merged**